sdk-actors/src/test/java/io/dapr/actors/runtime/ActorStatefulTest.java

```diff
@@ -0,0 +1,661 @@
+    ActorProxy proxy = newActorProxy();
+
+    // Zero is a magic input that will make method throw an exception.
+    proxy.invokeActorMethod("incrementAndGetCount", 0, int.class).block();
+  }
+
+  @Test(expected = IllegalStateException.class)
+  public void intTypeWithRuntimeException() {
+    ActorProxy proxy = newActorProxy();
+
+    proxy.invokeActorMethod("getCountButThrowsException", int.class).block();
+  }
+
+  @Test(expected = IllegalStateException.class)
+  public void actorRuntimeException() {
+    ActorProxy proxy = newActorProxy();
+
+    Assert.assertFalse(proxy.invokeActorMethod("hasMessage", Boolean.class).block());
+
+    proxy.invokeActorMethod("forceDuplicateException").block();
+  }
+
+  @Test(expected = IllegalCharsetNameException.class)
+  public void actorMethodException() {
+    ActorProxy proxy = newActorProxy();
+
+    Assert.assertFalse(proxy.invokeActorMethod("hasMessage", Boolean.class).block());
+
+    proxy.invokeActorMethod("throwsWithoutSaving").block();
+
+    Assert.assertFalse(proxy.invokeActorMethod("hasMessage", Boolean.class).block());
+  }
+
+  @Test
+  public void rollbackChanges() {
+    ActorProxy proxy = newActorProxy();
+
+    Assert.assertFalse(proxy.invokeActorMethod("hasMessage", Boolean.class).block());
+
+    // Runs a method that will add one message but fail because tries to add a second one.
+    proxy.invokeActorMethod("forceDuplicateException")
+      .onErrorResume(throwable -> Mono.empty())
+      .block();
+
+    // No message is set
+    Assert.assertFalse(proxy.invokeActorMethod("hasMessage", Boolean.class).block());
+  }
+
+  @Test
+  public void partialChanges() {
+    ActorProxy proxy = newActorProxy();
+
+    Assert.assertFalse(proxy.invokeActorMethod("hasMessage", Boolean.class).block());
+
+    // Runs a method that will add one message, commit but fail because tries to add a second one.
+    proxy.invokeActorMethod("forcePartialChange")
+      .onErrorResume(throwable -> Mono.empty())
+      .block();
+
+    // Message is set.
+    Assert.assertTrue(proxy.invokeActorMethod("hasMessage", Boolean.class).block());
+
+    // It is first message and not the second due to a save() in the middle but an exception in the end.
+    Assert.assertEquals("first message",
+      proxy.invokeActorMethod("getMessage", String.class).block());
+  }
+
+  private ActorProxy newActorProxy() {
+    ActorId actorId = newActorId();
+
+    // Mock daprClient for ActorProxy only, not for runtime.
+    DaprClient daprClient = mock(DaprClient.class);
+
+    when(daprClient.invokeActorMethod(
+      eq(context.getActorTypeInformation().getName()),
+      eq(actorId.toString()),
+      any(),
+      any()))
+      .thenAnswer(invocationOnMock ->
+        this.manager.invokeMethod(
+          new ActorId(invocationOnMock.getArgument(1, String.class)),
+          invocationOnMock.getArgument(2, String.class),
+          context.getActorSerializer().unwrapMethodResponse(

```

Comment:
This call to `unwrapMethodResponse` and the call a few lines after it to `wrapMethodRequest` read strangely, because at first it looks like you're calling something used to unwrap the response but on the *input* on line 612.  Then the `wrapMethodRequest` on 616 makes it look like you're wrapping the output using a method normally to wrap the *input*.

It looks like it's really filling in some functionality that is needed around the mocking, but a comment would help, because it almost looks "reversed".

Reply:
Ha. I plan to address this in the next PR when I will be adding custom serializer. This method should be called unwrapData() and the other one wrapData().

Reply:
That's fine.

Reply:
Renamed it in this PR.